### ISSUE TITLE
Rule Proposal: no unnecessary non-null assertion

### ISSUE THREAD

**Issue by bradzacher** [typescript-eslint/typescript-eslint#252]
We have no-non-null-assertion, but I would like to allow non-null assertions, but warn against using the non-null operator when there is no nullability. i.e.

```TS
function foo (a : number, b : number | undefined) {
  return (
    // should report
    a!.toString() +
    // is fine
    b!.toString()
  );
}
```

Would require renaming the rule.

**Comment by bradzacher** [comment]
Covered by [`no-unnecessary-type-assertion`](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-unnecessary-type-assertion.md)

We should update the docs to include an example so that it's clear that it supports it
